vllm/core/scheduler.py

```diff
@@ -1058,13 +1061,16 @@ def free_seq(self, seq: Sequence) -> None:
         self.block_manager.free(seq)
 
     def free_finished_seq_groups(self) -> None:
```

Comment:
@mzusman I actually don't get the current (before this PR) logic here.

If I understand correctly, the order of execution is `schedule` -> `get_and_reset_finished_requests_ids` -> model execution -> `free_finished_seq_groups`. Hence, the requests freed from `self.swapped` or `self.waiting` were not actually added to `finished_requests_ids` before the model execution. Is this the right order of execution?

Reply:
Ok, to my understanding,

1. Actually `finished_requests_ids` doesn't have to include the requests rejected in `self.waiting` since they are never passed into the model runner.
2. `finished_requests_ids` must include the requests rejected in `self.swapped` in the current step, which means the current code (before this PR) has a bug. However, the bug has not been noticed because this case is very rare.

Reply:
Yeah, That's correct, `finished_requests_ids` holds information that is lagged by 1 step. finished requests will always be released upon the next step.

AFAIU, `self.waiting` also includes preempted requests that got rescheduled, preempted requests did previously passed into the model runner and are already registered in the mamba cache. If those requests get aborted then we do add them to the `finished_requests_ids` and release them through [here](https://github.com/vllm-project/vllm/blob/135d8fefbf4242d4bf8c1a8b9f670047adc13b61/vllm/core/scheduler.py#L377) though. 
BTW Just to be sure, by requests rejected, do you mean aborted requests?

So actually, yeah, checking for finished requests in the `self.waiting` and `self.swapped` in `free_finished_seq_groups` is not necessary.

Reply:
@mzusman Thanks for the explanation! However, it's a bit unclear to me how it handles the preemption, both recompuation and swapping. Could you elaborate more on that?

Reply:
@WoosukKwon Sure! Upon recomputation we actually keep the Mamba cache as it is and do not evict it (since Mamba cache is quite small), the request id persists during the preemption therefore we can still use it's corresponding cache upon recomputation. And upon swapping - We do not handle swapping at the moment as it occurs fairly rarely and it's quite complicated to implement atm.

Reply:
@mzusman Thanks for the detailed explanation! It is super helpful to understand the implementation.

I think we should revisit this design decision in the near future. Currently, it's a bit confusing for code readers.